Bali/src/main/java/OfferList.java

```diff
@@ -0,0 +1,117 @@
+        offer.getLink() != null && offer.getLink().length() != 0 &&
+        offer.getVendor() != null && offer.getVendor().length() != 0 &&
+        offer.getHashTags() != null && offer.getHashTags().size() != 0 && offer.getHashTags().size() <= 7 &&
+        offer.getValidUntil() != null && offer.getValidUntil().length() != 0 &&
+        offer.getPhotoLink() != null && offer.getPhotoLink().length() != 0 &&
+        offer.getDiscount() != null && offer.getDiscount().length() != 0 &&
+        offer.getReviews() != null && offer.getReviews().size() != 0){
+            return true;
+        }
+        return false;
+    }
+
+    public boolean addOffer(Offer offer){
+        if(OfferList.validateOffer(offer) && getOffer(offer.getId()) == null){
+            offers.add(offer);
+            return true;
+        }
+        return false;
+    }
+
+    public boolean editOffer(String id, Offer offer){
+        if(offer.getId() != null || offer.getVendor() != null || offer.getCreatedAt() != null){
+            return false;
+        }
+        Offer newOffer = getOffer(id);
+        if(offer.getDescription() != null){
+            newOffer.setDescription(offer.getDescription());
+        }
+        if(offer.getLink() != null){
+            newOffer.setLink(offer.getLink());
+        }
+        if(offer.getHashTags() != null){
+            newOffer.setHashTags(offer.getHashTags());
+        }
+        if(offer.getValidUntil() != null){
+            newOffer.setValidUntil(offer.getValidUntil());
+        }
+        if(offer.getPhotoLink() != null){
+            newOffer.setPhotoLink(offer.getPhotoLink());
+        }
+        if(offer.getDiscount() != null){
+            newOffer.setDiscount(offer.getDiscount());
+        }
+        if(offer.getReviews() != null){
+            newOffer.setReviews(offer.getReviews());
+        }
+        if(offer.getRating() != 0){
+            newOffer.setRating(offer.getRating());
+        }
+        return true;
+    }
+
+    public boolean removeOffer(String id){
+        if(getOffer(id) != null){
+            offers.remove(getOffer(id));
+            return true;
+        }
+        return false;
+    }
+
+    public List<Offer> getPage(int skip, int top, OfferFilter offerFilter){
+        List<Offer> bufOfferList = new ArrayList<>();
+        if(skip >= offers.size()){
+            return null;
+        }
+        int amountOffers = offers.size();
+        for(int i = skip; i < amountOffers && top > 0; i++){
+            if(offers.get(i).getVendor().equals(offerFilter.getVendor()) ||
+                    offers.get(i).getValidUntil().equals(offerFilter.getValidUntil()) ||
+                    offers.get(i).getHashTags().containsAll(offerFilter.getHashTags())) {
+                bufOfferList.add(offers.get(i));
+                top--;
+            }
+        }
+        return bufOfferList.stream().sorted((offer1, offer2) -> {
+            SimpleDateFormat formatter = new SimpleDateFormat("dd.MM.yyyy", Locale.ENGLISH);
```

Comment:
It's way better to encapsulate string parsing into `Offer` itself. Then you would go without a multiline lambda (which is not pretty)